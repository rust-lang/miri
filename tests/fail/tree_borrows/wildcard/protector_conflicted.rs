//@compile-flags: -Zmiri-tree-borrows -Zmiri-permissive-provenance

#[allow(unused_variables)]
pub fn main() {
    let mut x: u32 = 42;

    let ptr_base = &mut x as *mut u32;
    let ref1 = unsafe { &mut *ptr_base };
    let ref2 = unsafe { &mut *ptr_base };

    let protect = |arg: &mut u32| {
        // expose arg
        let int = arg as *mut u32 as usize;
        let wild = int as *mut u32;

        // foreign read to arg marks it as conflicted making child_writes UB while its protected
        let _x = *ref2;

        // UB because it tries to write through arg
        unsafe { *wild = 4 }; //~ ERROR: /write access through wildcard at .* is forbidden/
    };

    protect(ref1);
}
