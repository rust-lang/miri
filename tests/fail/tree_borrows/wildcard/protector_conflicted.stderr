error: Undefined Behavior: write access through wildcard at ALLOC[0x0] is forbidden
  --> tests/fail/tree_borrows/wildcard/protector_conflicted.rs:LL:CC
   |
LL |         unsafe { *wild = 4 };
   |                  ^^^^^^^^^ Undefined Behavior occurred here
   |
   = help: this indicates a potential bug in the program: it performed an invalid operation, but the Tree Borrows rules it violated are still experimental
   = help: see https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/tree-borrows.md for further information
   = help: there are no exposed references who have write access permissions to this location
   = note: BACKTRACE:
   = note: inside closure at tests/fail/tree_borrows/wildcard/protector_conflicted.rs:LL:CC
note: inside `main`
  --> tests/fail/tree_borrows/wildcard/protector_conflicted.rs:LL:CC
   |
LL |     protect(ref1);
   |     ^^^^^^^^^^^^^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to 1 previous error

